### ISSUE TITLE
FP32 training for sd3 controlnet

### ISSUE THREAD

**Issue by xduzhangjiayu** [huggingface/diffusers#9560]
Hi,
I have been use `examples\controlnet\train_controlnet_sd3.py` for controlnet training for a while, and I have some confusion and would like your advice

1. In the line 1097:
`vae.to(accelerator.device, dtype=torch.float32)`
It seems we should use fp32 for VAE, but as far as I know, SD3 currently has no fp32 checkpoints, so does it really work if we populate fp16 into fp32?

2. Before running the train script, `accelerate config` can specify whether to use mixed precision or not, since SD3 only has fp16 checkpoint at present, I don't know how to choose this option, whether to choose 'fp16' or 'no'.

Really appreciate your advice!
@sayakpaul @DavyMorgan

**Comment by DavyMorgan** [comment]
Hi, here we use fp32 to prevent numerical instabilities, and you can check [this](https://github.com/huggingface/diffusers/pull/9249#discussion_r1742878371).
Also, in my experiments, I choose 'no' during `accelerate config`.

**Comment by xduzhangjiayu** [comment]
Thanks for the reply!
The SD3 base model you used is fp32?

**Comment by tolgacangoz** [comment]
[I have also wondered this :thinking:.](https://huggingface.co/stabilityai/stable-diffusion-3-medium-diffusers/discussions/48)

**Comment by xduzhangjiayu** [comment]
As author of ControlNet mentioned here, train ControlNet with FP32 will not work well. 
https://github.com/lllyasviel/ControlNet/issues/265

**Comment by DavyMorgan** [comment]
@sayakpaul If I understand it correctly, we cast the fp16 weight to fp32 to prevent numerical instabilities (you are right that SD3 currently has no fp32 checkpoints).
Meanwhile, it seems that training ControlNet with FP16 instead of FP32 will not work well from https://github.com/lllyasviel/ControlNet/issues/265#issuecomment-1466354258

**Comment by xduzhangjiayu** [comment]
> @sayakpaul If I understand it correctly, we cast the fp16 weight to fp32 to prevent numerical instabilities (SD3 currently has no fp32 checkpoints). @xduzhangjiayu Meanwhile, it seems that training ControlNet with FP16 rather than FP32 will not work well from [lllyasviel/ControlNet#265 (comment)](https://github.com/lllyasviel/ControlNet/issues/265#issuecomment-1466354258)

In my experiment, the loss began to oscillate continuously after ~10,000 steps without convergence, and the result image was also fuzzy and blocky but semantically correct. I am not sure whether this is related to the fp16 training.

**Comment by sayakpaul** [comment]
@xduzhangjiayu thanks for opening the issue. 

VAE upcasting was something as needed for DreamBooth LoRA trainings especially using the FP16 mixed-precision training. For BF16, it may not be a problem. This is analogous to why https://huggingface.co/madebyollin/sdxl-vae-fp16-fix was needed for SDXL FP16 fine-tuning. 

> Before running the train script, accelerate config can specify whether to use mixed precision or not, since SD3 only has fp16 checkpoint at present, I don't know how to choose this option, whether to choose 'fp16' or 'no'.

You can choose "fp16" or "bf16" depending on your hardware. But if you specify `--mixed_precision="fp16"` when launching training, it should work.

**Comment by sayakpaul** [comment]
I don't think upcasting is invalid. We _always_ upcast the VAE for numerical stabilities. But we could maybe add a CLI option so disable it. Would you maybe like to open a PR for that?

**Comment by xduzhangjiayu** [comment]
Sure, I would like do that.

**Comment by sayakpaul** [comment]
https://github.com/huggingface/diffusers/pull/9601
